File: 09_Generics/02_variance/java/ArrayVariance.java
class A {}
class B extends A {}
class C extends B {}

public class ArrayVariance {

	private static int processAs(A[] aas) {
		int count = 0;
		for (int i = 0; i < aas.length; ++i) {
			count += 1;
			aas[i] = new C(); // allowed?
			aas[i] = new A(); // allowed?
		}
		return count;
	}

	private static int processCs(C[] acs) {
		int count = 0;
		for (int i = 0; i < acs.length; ++i) {
			count += 1;
		}
		return count;
	}

	public static void main(String[] args) {
			
		B[] bs = new B[] { new B(), new B(), new B(), new C() };

		int na = processAs(bs);
		//int nc = processCs(bs);
		
		System.out.println(na);
		//System.out.println(nc);
	}
}

File: 09_Generics/02_variance/kotlin/ArrayVariance.kt
import kotlin.reflect.*
import kotlin.reflect.full.*

open class A
open class B : A()
open class C : B()

/*
fun processAs(aas: Array<A>): Int {
    var count = 0
    for (i in aas.indices) {
        count += 1
    }
    return count
}
*/
/*
fun <T : A> processAs(aas: Array<T>): Int {
    var count = 0
    for (i in aas.indices) {
        count += 1
        aas[i] = A()
    }
    return count
}
*/

inline fun <reified T : A> processAs(aas: Array<T>): Int {
    var count = 0
    for (i in aas.indices) {
        count += 1
        aas[i] = T::class.createInstance()
    }
    return count
}

fun processCs(acs: Array<C>): Int {
    var count = 0
    for (i in acs.indices) {
        count += 1
    }
    return count
}

fun main() {
    val bs = arrayOf(B(), B(), B(), C())
    println(bs[3]::class)

    val na = processAs(bs)
    //val nc = processCs(bs)

    println(na)
    //println(nc)
		
    println(bs[3]::class)
}

